torch/optim/_multi_tensor/sgd.py

```diff
@@ -0,0 +1,140 @@
+                v_{t+1} & = \mu * v_{t} + \text{lr} * g_{t+1}, \\
+                p_{t+1} & = p_{t} - v_{t+1}.
+            \end{aligned}
+
+        The Nesterov version is analogously modified.
+    """
+
+    def __init__(self, params, lr=required, momentum=0, dampening=0,
+                 weight_decay=0, nesterov=False):
+        if lr is not required and lr < 0.0:
+            raise ValueError("Invalid learning rate: {}".format(lr))
+        if momentum < 0.0:
+            raise ValueError("Invalid momentum value: {}".format(momentum))
+        if weight_decay < 0.0:
+            raise ValueError("Invalid weight_decay value: {}".format(weight_decay))
+
+        defaults = dict(lr=lr, momentum=momentum, dampening=dampening,
+                        weight_decay=weight_decay, nesterov=nesterov)
+        if nesterov and (momentum <= 0 or dampening != 0):
+            raise ValueError("Nesterov momentum requires a momentum and zero dampening")
+        super(SGD, self).__init__(params, defaults)
+
+    def __setstate__(self, state):
+        super(SGD, self).__setstate__(state)
+        for group in self.param_groups:
+            group.setdefault('nesterov', False)
+
+    @torch.no_grad()
+    def step(self, closure=None):
+        """Performs a single optimization step.
+
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            with torch.enable_grad():
+                loss = closure()
+
+        for group in self.param_groups:
+            weight_decay = group['weight_decay']
+            momentum = group['momentum']
+            dampening = group['dampening']
+            nesterov = group['nesterov']
+
+            grads = []
+            params_with_grad = []
+            states = []
+            has_sparse_grad = False
+
+            for p in group['params']:
+                if p.grad is not None:
+                    grads.append(p.grad)
+                    params_with_grad.append(p)
+                    states.append(self.state[p])
+
+                    if p.grad.is_sparse:
+                        has_sparse_grad = True
+
+                        if momentum != 0: 
+                            raise RuntimeError('SGD does not support momentum for sparse gradients')
+
+            if grads == []:
+                return loss
+
+            if weight_decay != 0:
+                grads = torch._foreach_add(grads, params_with_grad, alpha=weight_decay)
+
+            if momentum != 0:
+                bufs = []
+                for i in range(len(states)):
+                    if 'momentum_buffer' not in states[i]:
+                        buf = states[i]['momentum_buffer'] = torch.clone(grads[i]).detach()
+                    else:
+                        buf = states[i]['momentum_buffer']
+                        buf.mul_(momentum).add_(grads[i], alpha=1 - dampening)
```

Comment:
This should also be repackaged as an MTA calculation, it's dominating the runtime in my tests (see picture).

Reply:
As @ngimel pointed out, many optimizers have lingering per-tensor loops that should be repackaged as MTA.  This can be done over time.  It would be great for Adam and SGD, at least, to fully avoid cpu-bound loops over tensors in 1.7.